Replace deprecated H5part with H5hut

biddisco opened the pull request:
H5part was discontinued many years ago, but the code was copied into the source tree of sphexa and although it works fine when built as an external as part of the project, it ought to be replaced with the newer H5hut library that contains all the functions of H5part, but with additional block data support.
This PR removes H5part, and adds H5hut instead, but not built as part of the source tree. H5hut must be installed on the system using (for example)
```
spack install h5hut@2:
spack install h5hut@master
```

Some small changes to the API for access to numparticles and particle data/attributes are reflected in the PR and the access to type safe attributes has been rewritten slightly.

Note - this PR relies on H5hut merge request https://gitlab.psi.ch/H5hut/src/-/merge_requests/7 and should not be merged before the int8 changes have been accepted into H5hut.
: update - these changes have been merged

:update: Note - users of sphexa will NOT need to change their build environment to add H5hut, cmake fetchcontent code has been added to ensure that h5hut is downloaded and compiled in place

biddisco committed:
Remove h5part source from sphexa tree

biddisco commented:
int8 PR has been merged into H5hut master.

biddisco committed:
Rename vars/files using H5PART/H5Part/H5Hut to 'H5hut' for naming consistency

sekelle commented on main/src/io/ifile_io_hdf5.cpp:
```diff
@@ -179,13 +185,13 @@ class H5PartReader final : public IFileReader
         closeStep();
         pathStep_ = path;
-        h5File_   = fileutils::openH5Part(path, H5PART_READ | H5PART_VFD_MPIIO_IND, comm_);
+        h5File_   = fileutils::openH5Part(path, H5_O_RDWR, comm_);
```
Have you checked that this is not doing collective MPI I/O?

sekelle commented on main/src/io/h5part_wrapper.hpp:
```diff
@@ -198,163 +237,61 @@ void readAttribute(ExtractType* attr, int attrSizeBuf, int attrIndex, Reader&& r
-    h5_file = H5PartOpenFileParallel(h5_fname, mode, comm);
+#ifdef H5_HAVE_PARALLEL
+    h5_prop_t prop = H5CreateFileProp();
+    H5SetPropFileMPIOCollective(prop, &comm);
```
Ah, it is collective. Independent performs better, can you change it?

jfavre commented:
Nice effort, but I wonder if any discussion has taken place about the merits of h5hut which AFAIK, is not astro-, or fluid-specific, compared to, for example Gagdet, or some much more recent and physics oriented format such as openPMD for particle physics. H5part had such poor support for meta data convention...I hope h5hut does.
And btw, who will support an h5hut ParaView reader? N.B. Gadget and openPMD already have ParaView readers.

sekelle commented:
I believe the meta-data convention in h5hut is still the same, i.e the resulting HDF5 output files will have the same structure as what we've been writing with H5Part.

`openPMD` seems very interesting.  It's also a thin layer over HDF5, enforcing a certain data layout and some common sense attributes. The option to use ADIOS or JSON instead is nice.

biddisco committed:
Add option to control independent/collective IO for HDF5

biddisco commented:
> Ah, it is collective. Independent performs better, can you change it?

Pushed a commit with an option to switch between independent and collective, with independent as default. 
(NB. rebased branch onto latest master)

biddisco commented:
> Nice effort, but I wonder if any discussion has taken place about the merits of h5hut which AFAIK, is not astro-, or fluid-specific, compared to, for example Gagdet, or some much more recent and physics oriented format such as openPMD for particle physics. H5part had such poor support for meta data convention...I hope h5hut does. And btw, who will support an h5hut ParaView reader? N.B. Gadget and openPMD already have ParaView readers.

As sebastien mentioned, this does not change the IO at all, only upgrades the lib version used. The h5part support in sphexa was using an old version of the code that I have upgraded to h5hut - exact same binary hdf5 file contets, no change to readers.

Adding support for something more widely used like gadget or openPMD would be a very good idea -  they most likely have more tools available for processing them. I'm sure when someone needs extra functioanlity from sphexa files, they can/will ask for it ;)

jfavre commented:
in a recent past, I ran into what seems to be a problem with ParaView v5.13 (earlier versions not tested), when running in parallel, over multiple timesteps, particles gradually disappear. @ChristopherBignamini  and I have seen that last week, but because of time pressure, I have not had time to document it.
I feel that before doing the upgrade to h5hut, we should build a test that demonstrates proper parallel, time-dependent. When that test passes, we can upgrade and test again.

jgphpc committed:
add yaml code for ci, and update for debug/release

jgphpc commented:
https://cicd-ext-mw.cscs.ch/ci/pipeline/results/4264416954089684/31431274/1722271041?iid=2521&type=gitlab

jfavre commented:
I have created a Python script to generate an example H5part and have created a ParaView issue to say that the H5part reader in ParaView does not work in parallel.
https://gitlab.kitware.com/paraview/paraview/-/issues/22918

jgphpc commented:
waiting for: 
1/ https://gitlab.kitware.com/paraview/paraview/-/issues/21119 to be solved
2/ ci to be ready (GB)

biddisco commented:
Recently the git repo for h5hut was moved from one PSI server to another - this breaks the spack package for h5hut so that `spack install` fails for it. Unfortunately, the new git repo is private and so even if we change the repoin spack, it still won't work for general users. I have asked for the repo to be moved to github or somewhere more public and am waiting for that to happen. until then, probably it is best to not merge this PR.

Note that the fact that paraview has a buggy h5part reader is not directly relevant to this PR - the actual hdf5 code that writes data is unchanged in h5hut from h5part and this PR only changes the library used and the source location. 
It is annoying that the patch for paraview won't be applied any time soon and one solution is for me to update my h5part (now h5hut) plugin and add that to the paraview uenv and also make it available on github as source for anyone to compile and use locally.

biddisco commented:
Please do not make commits with a log message such as "fix" (eg https://github.com/sphexa-org/sphexa/pull/458/commits/c4fbf7f48cb324e86f48d3650894973ea128a4fe or https://github.com/sphexa-org/sphexa/pull/458/commits/036ec81e3e0ed5cdaf11dd5cec690f899e93bc74 and 7 more at first glance). A git commit message should have something meaningful that tells others what it fixes without them having to click and read additional info such as the diff.
if necessary squash them all into one commit and use a single message.

biddisco commented:
fyi : h5hut spack package has been updated with new github repo location + other fix https://github.com/spack/spack-packages/pull/598

biddisco committed:
Replace deprecated h5part with newer H5hut usage and simplify some calls

h5part was copied as a snapshot into sphexa but was out of date and this
upgrades the h5part wrapper code to use h5hut instead.

H5Hut supports the same particle datasets as h5part, but also adds
support for volumentric fields and has a slightly more consistent use of
hdf5 datatypes throughout

This patch requires the use of (to be released) h5hut version 2.0.0
or the master branch from git

Note:
NumParticles/Steps etc have changed from unsigned to signed variables

biddisco committed:
Use FetchContent to get H5hut from git if required

We use FetchContent with a version check that allows the user to either
find_package and get a system installed H5hut, or if the version is
prefixed by "git.xxx" then fetch the branch/sha/tag from git and build
in tree on demand

Note that Fetchcontent with find_package support reuires cmake 3.24

biddisco commented:
This branch has been rebased onto develop with conflicts fixed and tests checked locally, in addition I have added a fetchcontent for h5hut to avoid needing a system install of it (but you can still use that if you want).
Note that cmake minimum version bumps to 3.24 to support fetchcontent with find-package support

biddisco committed:
Merge branch 'develop' into h5hut

jgphpc commented:
For info: https://github.com/sphexa-org/sphexa/pull/531 -> main/src/io/ifile_io_hdf5.cpp

jgphpc committed:
Fix conflict (1)

jgphpc committed:
Fix conflit (2)

jgphpc committed:
Fix conflict (3)

jgphpc committed:
Fix conflict (4)

jgphpc committed:
Merge branch 'develop' into h5hut

jgphpc commented:
cscs-ci run

jgphpc commented:
I am failing to build:
```
[ 11%] Building CXX object main/src/io/CMakeFiles/io.dir/ifile_io_hdf5.cpp.o
cd /tmp/piccinal/spack-stage/spack-stage-sphexa-7b3037e8-gfbi6ujtlhwwussv6nk4yoho2eb2rbvc/spack-build-gfbi6uj/main/src/io
&&
/user-environment/linux-neoverse_v2/cray-mpich-8.1.32-nf6wcjedmweoe2xygtmwickmsdqevt7w/bin/mpicxx
-I/tmp/piccinal/spack-stage/spack-stage-sphexa-7b3037e8-gfbi6ujtlhwwussv6nk4yoho2eb2rbvc/spack-src/domain/include
-g -std=gnu++20 -MD -MT main/src/io/CMakeFiles/io.dir/ifile_io_hdf5.cpp.o -MF
CMakeFiles/io.dir/ifile_io_hdf5.cpp.o.d -o
CMakeFiles/io.dir/ifile_io_hdf5.cpp.o -c
/tmp/piccinal/spack-stage/spack-stage-sphexa-7b3037e8-gfbi6ujtlhwwussv6nk4yoho2eb2rbvc/spack-src/main/src/io/ifile_io_hdf5.cpp

In file included from /user-environment/linux-neoverse_v2/gcc-14.2.0-kbd4rtpzhot3mo3djvfjz6nr5j2yvsjj/lib/gcc/aarch64-unknown-linux-gnu/14.2.0/../../../../include/c++/14.2.0/bits/shared_ptr_base.h:59,
                 from /user-environment/linux-neoverse_v2/gcc-14.2.0-kbd4rtpzhot3mo3djvfjz6nr5j2yvsjj/lib/gcc/aarch64-unknown-linux-gnu/14.2.0/../../../../include/c++/14.2.0/bits/shared_ptr.h:53,
                 from /user-environment/linux-neoverse_v2/gcc-14.2.0-kbd4rtpzhot3mo3djvfjz6nr5j2yvsjj/lib/gcc/aarch64-unknown-linux-gnu/14.2.0/../../../../include/c++/14.2.0/bits/fs_path.h:46,
                 from /user-environment/linux-neoverse_v2/gcc-14.2.0-kbd4rtpzhot3mo3djvfjz6nr5j2yvsjj/lib/gcc/aarch64-unknown-linux-gnu/14.2.0/../../../../include/c++/14.2.0/filesystem:52,
                 from /tmp/piccinal/spack-stage/spack-stage-sphexa-7b3037e8-gfbi6ujtlhwwussv6nk4yoho2eb2rbvc/spack-src/main/src/io/ifile_io_hdf5.cpp:33:

/user-environment/linux-neoverse_v2/gcc-14.2.0-kbd4rtpzhot3mo3djvfjz6nr5j2yvsjj/lib/gcc/aarch64-unknown-linux-gnu/14.2.0/../../../../include/c++/14.2.0/bits/unique_ptr.h:
In instantiation of 'std::__detail::__unique_ptr_t<_Tp>
std::make_unique(_Args&& ...) [with _Tp = sphexa::UnimplementedReader; _Args =
{}; __detail::__unique_ptr_t<_Tp> =
__detail::__unique_ptr_t<sphexa::UnimplementedReader>]':

/tmp/piccinal/spack-stage/spack-stage-sphexa-7b3037e8-gfbi6ujtlhwwussv6nk4yoho2eb2rbvc/spack-src/main/src/io/ifile_io_hdf5.cpp:474:103:
required from here

  474 | std::unique_ptr<IFileReader> makeH5PartReader(MPI_Comm) { return std::make_unique<UnimplementedReader>(); }



    /user-environment/linux-neoverse_v2/gcc-14.2.0-kbd4rtpzhot3mo3djvfjz6nr5j2yvsjj/lib/gcc/aarch64-unknown-linux-gnu/14.2.0/../../../../include/c++/14.2.0/b
            its/unique_ptr.h:1076:30: error: invalid new-expression of abstract class type 'sphexa::UnimplementedReader'

```

biddisco committed:
Fix missing pure virtual file/step attribute override in UnimplementedReader

biddisco committed:
Fix H5PART->H5HUT variables to be consistent with capitalized H5HUT

biddisco committed:
Fix H5 sequential writer to use new H5hut syntax, add fileAttrib test (c.f. stepAttrib test)

jgphpc reviewed (approved):
[lgtm](https://gitlab.com/cscs-ci/ci-testing/webhook-ci/mirrors/44692846847247/4264416954089684/-/pipelines/2017833826) (however I did not check the content of the h5 files)

sekelle reviewed (changes requested):
The main checkpointing and restore works. However, there are two issues:

* The sequential HDF5 writer does not work and hangs with multiple ranks trying write the profiling output
* The parallel HDF5 writer reports the local instead of global particle count when the `globalNumParticle()` function is called.

```bash
OMP_NUM_THREADS=64 srun -N1 -c72 --ntasks-per-node=4 --gpus-per-task=1 ./sphexa-cuda --init sedov -s 2 -n 500 --profile
```

biddisco committed:
Add a dummy communicator of just one rank to the sequential HDF5 particle writer

biddisco committed:
Fix Sequential hdf5 writer to write data from all ranks, add test

biddisco committed:
Make sequential H5Part writer a subclass of primary H5Part writer

biddisco committed:
Always check writer global count when testing IO

biddisco committed:
H5hut only reports numParticles in view, so add AllReduce for the total

biddisco committed:
Merge remote-tracking branch 'origin/develop' into h5hut

sekelle reviewed (approved):
All issues are fixed now.
